csrc/mamba/causal_conv1d/causal_conv1d.cu

```diff
@@ -0,0 +1,930 @@
+    TORCH_CHECK(seq_pos_idx.is_cuda());
+    TORCH_CHECK(seq_pos_idx.is_contiguous());
+    CHECK_SHAPE(seq_pos_idx, batch_size, seqlen);
+  }
+  at::Tensor out = torch::empty_like(x);
+
+  ConvParamsBase params;
+  set_conv_params_fwd(params, batch_size, dim, seqlen, width, x, weight, out,
+                      bias_.has_value() ? bias_.value().data_ptr() : nullptr,
+                      silu_activation);
+
+  if (seq_idx_.has_value()) {
+    params.seq_idx_ptr = seq_idx_.value().data_ptr();
+  } else {
+    params.seq_idx_ptr = nullptr;
+  }
+
+  if (seq_pos_idx_.has_value()) {
+    params.seq_pos_idx_ptr = seq_pos_idx_.value().data_ptr();
+  } else {
+    params.seq_pos_idx_ptr = nullptr;
+  }
+  if (initial_states_.has_value()) {
+    TORCH_CHECK(is_channel_last,
+                "initial_states is only supported for channel last layout");
+    auto initial_states = initial_states_.value();
+    TORCH_CHECK(initial_states.scalar_type() == input_type);
+    TORCH_CHECK(initial_states.is_cuda());
+    CHECK_SHAPE(initial_states, batch_size, dim, width - 1);
+    TORCH_CHECK(initial_states.stride(1) == 1);
+    params.initial_states_ptr = initial_states.data_ptr();
+    params.initial_states_batch_stride = initial_states.stride(0);
+    params.initial_states_c_stride = initial_states.stride(1);
+    params.initial_states_l_stride = initial_states.stride(2);
+  } else {
+    params.initial_states_ptr = nullptr;
+  }
+
+  if (final_states_out_.has_value()) {
+    TORCH_CHECK(is_channel_last,
+                "final_states is only supported for channel last layout");
+    auto final_states = final_states_out_.value();
+    TORCH_CHECK(final_states.scalar_type() == input_type);
+    TORCH_CHECK(final_states.is_cuda());
+    CHECK_SHAPE(final_states, batch_size, dim, width - 1);
+    TORCH_CHECK(final_states.stride(1) == 1);
+    params.final_states_ptr = final_states.data_ptr();
+    params.final_states_batch_stride = final_states.stride(0);
+    params.final_states_c_stride = final_states.stride(1);
+    params.final_states_l_stride = final_states.stride(2);
+  } else {
+    params.final_states_ptr = nullptr;
+  }
+
+  // Otherwise the kernel will be launched from cuda:0 device
+  // Cast to char to avoid compiler warning about narrowing
+  at::cuda::CUDAGuard device_guard{(char)x.get_device()};
+  auto stream = at::cuda::getCurrentCUDAStream().stream();
+  DISPATCH_ITYPE_FLOAT_AND_HALF_AND_BF16(
+      x.scalar_type(), "causal_conv1d_fwd", [&] {
+        DISPATCH_WTYPE_FLOAT_AND_HALF_AND_BF16(
+            weight.scalar_type(), "causal_conv1d_fwd", [&] {
+              if (!is_channel_last) {
+                causal_conv1d_fwd_cuda<input_t, weight_t>(params, stream);
+              } else {
+                causal_conv1d_channellast_fwd_cuda<input_t, weight_t>(params,
+                                                                      stream);
+              }
+            });
+      });
```

Comment:
Looks like this is generating every type combination. If not necessary, I think we should pare this down.

Reply:
Aligned it to use the same dtype for both the itype and wtype